src/project.ts

```diff
@@ -167,4 +167,84 @@ async function addEditorSettingsButton () {
 	}
 }
 
-export { addProgramInfo, keyboardShortcuts, addEditorSettingsButton, checkHiddenOrDeleted };
+function saveNewScratchpadToLocalStorage () {
+	let aceEditor;
+	let checkIfEditorIsReadyInterval = setInterval(function () {
+		aceEditor = document.getElementsByClassName("scratchpad-ace-editor")[0]?.env?.editor;
+
+		if (aceEditor !== undefined) {
+			// get the type of the program
+			const programType = window.location.href.split("/")[5];
+			if (!["pjs", "webpage", "sql"].includes(programType)) {
+				clearInterval(checkIfEditorIsReadyInterval); // we can stop checking if the editor is ready
+				return;
+			}
+
+			// get user kaid
+			let userKaid;
+			for (const key in window.__APOLLO_CLIENT__.cache.data.data) {
+				if (key.startsWith("User:kaid_")) {
+					userKaid = key.slice("User:".length, key.indexOf("."));
+				}
+			}
+			
+			const localStorageKey = "ka:4:cs-scratchpad-new" + userKaid + "-" + programType;
+			
+			let scratchpadObject;
+
+			// get previous code and write it to the ace editor
+			scratchpadObject = localStorage.getItem(localStorageKey);
+			if (typeof scratchpadObject === "string" && scratchpadObject.length > 0) {
+				try {
+					scratchpadObject = JSON.parse(scratchpadObject);
+					if (scratchpadObject.scratchpad.revision.code.length > 0) {
+						aceEditor.setValue(scratchpadObject.scratchpad.revision.code);
+					}
+				} catch (e) {
+					console.log("Failed to load scratchpad from local storage");
+				}
+			}
+			
+			// save user code to local storage
+			function saveEditorCode () {
+				scratchpadObject = {
+					"cursor": {
+						"row": 0,
+						"column": 0
+					},
+					"scratchpad": {
+						"title": "New " + (programType === "pjs" ? "program" : "webpage"),
+						"translatedTitle": "New " + (programType === "pjs" ? "program" : "webpage"),
+						"category": null,
+						"difficulty": null,
+						"userAuthoredContentType": programType,
+						"revision": {
+							"code": aceEditor.getValue(),
+							"created": new Date().toISOString()
+						},
+						"trustedRevision": {
+							"created": new Date().toISOString()
+						}
+					}
+				};
+				localStorage.setItem(localStorageKey, JSON.stringify(scratchpadObject));
+			}
+			window.addEventListener("beforeunload", saveEditorCode); // save code when the user exits the page
```

Comment:
Make sure to avoid saving the code to localStorage if we aren't currently on a new scratchpad page.  Going back to the program parameter I mentioned before, you could check if the id is null or not before saving.